addons/stock/models/stock_picking.py

```diff
@@ -282,13 +282,11 @@ def _compute_use_existing_lots(self):
                 picking_type.use_existing_lots = True
 
     @api.model
-    def _name_search(self, name, domain=None, operator='ilike', limit=None, order=None):
-        # Try to reverse the `display_name` structure
-        parts = name.split(': ')
-        if len(parts) == 2:
-            name_domain = [('warehouse_id.name', operator, parts[0]), ('name', operator, parts[1])]
-            return self._search(expression.AND([name_domain, domain]), limit=limit, order=order)
-        return super()._name_search(name, domain, operator, limit, order)
+    def _search_display_name(self, operator, value):
+        parts = isinstance(value, str) and value.split(': ')
+        if parts and len(parts) == 2:
+            return ['&', ('warehouse_id.name', operator, parts[0]), ('name', operator, parts[1])]
+        return super()._search_display_name(operator, value)
```

Comment:
This gave me an idea.
Quite obviously the domain inside the `if` statement is incorrect when `operator` is negative.
And I see that most `_search_display_name` methods keep on doing the same logic to handle the negative operator case.

What about ensuring that `operator` is always positive?
I mean, instead of replacing `[('display_name', operator, value)]` by
```py
domain = model._search_display_name(operator, value)
```
we could factor out the problem of the negation as
```py
if operator in NEGATIVE_TERM_OPERATORS:
    positive_operator = TERM_OPERATORS_NEGATION[operator]
    positive_domain = model._search_display_name(positive_operator, value)
    domain = expression.NOT(positive_domain)
else:
    domain = model._search_display_name(operator, value)
```

Reply:
Every domain is not yet complemented correctly (#171291 and [(x) IS NOT TRUE](https://github.com/odoo/odoo/pull/174829/files#diff-fa4d9268d6e65e19aebec81c46038f0e496b91142588ed4d1c1bce7ff2338f2cR963)).

The idea is good: we can implement this after refactoring the domains where we would always call _search methods with positive operators. I logged task-4150178